File: javascript/modifying_sequence_spec.js
var should = require('chai').should();
var Algos = require('./modifying_sequence');

var Ary = function() {
  return [1, 2, 3, 4, 5];
};

var Hash = function() {
  return {'a': 1, 'b': 2};
};

describe('Modifying sequence algorithms', function(){

  it('copy', function(){
    var hash = {};
    Algos.copy(Hash(), hash);
    JSON.stringify(hash).should.be.equal(JSON.stringify({'a': 1, 'b': 2}));
  });

  it('copy_if', function(){
    var hash = {};
    Algos.copy_if(Hash(), hash, function(e){
      return e === 1;
    });
    JSON.stringify(hash).should.be.equal(JSON.stringify({'a': 1}));
  });

  it('copy_n', function(){
    var ary = new Array(3);
    Algos.copy_n(ary, 3, 3);
    JSON.stringify(ary).should.be.equal(JSON.stringify([3, 3, 3]));
  });

  it('copy_backward', function(){
    var ary = new Array(3);
    Algos.copy_backward(ary, Ary()).should.be.equal(-1);
    ary = new Array(5);
    Algos.copy_backward(ary, Ary()).should.be.equal(0);
    JSON.stringify(ary).should.be.equal(JSON.stringify([1, 2, 3, 4, 5]));
  });

  // move* is implemented the same as copy*
  it('move', function(){
    var hash = {};
    Algos.move(Hash(), hash);
    JSON.stringify(hash).should.be.equal(JSON.stringify({'a': 1, 'b': 2}));
  });

  it('move_backward', function(){
    var ary = new Array(3);
    Algos.move_backward(ary, Ary()).should.be.equal(-1);
    ary = new Array(5);
    Algos.move_backward(ary, Ary()).should.be.equal(0);
    JSON.stringify(ary).should.be.equal(JSON.stringify([1, 2, 3, 4, 5]));
  });

  it('fill', function(){
    var ary = new Array(3);
    Algos.fill(ary, 3);
    JSON.stringify(ary).should.be.equal(JSON.stringify([3, 3, 3]));
  });

  it('fill_n', function(){
    var ary = new Array(3);
    Algos.fill_n(ary, 2, 3);
    JSON.stringify(ary).should.be.equal(JSON.stringify([3, 3, null]));
  });

  it('transform', function(){
    var emptyAry = [];
    JSON.stringify(Algos.transform(Ary(), function(e){
      emptyAry.push(e + 1);
    })).should.be.equal(JSON.stringify(Ary()));
    JSON.stringify(emptyAry).should.be.equal(JSON.stringify([2, 3, 4, 5, 6]));
  });

  it('generate', function(){
    var ary = new Array(3);
    Algos.generate(ary, function(){
      return 3;
    });
    JSON.stringify(ary).should.be.equal(JSON.stringify([3, 3, 3]));
  });

  it('generate_n', function(){
    var ary = new Array(3);
    Algos.generate_n(ary, 2, function(){
      return 3;
    });
    JSON.stringify(ary).should.be.equal(JSON.stringify([3, 3, null]));
  });

  it('remove', function(){
    var ary = [1, 2, 3];
    Algos.remove(ary, 3);
    JSON.stringify(ary).should.be.equal(JSON.stringify([1, 2]));
  });

  it('remove_if', function(){
    var ary = [1, 2, 3];
    Algos.remove_if(ary, function(e){
      return e < 2;
    });
    JSON.stringify(ary).should.be.equal(JSON.stringify([2, 3]));
    var hash = {'a': 1, 'b': 2};
    Algos.remove_if(hash, function(e){
      return e != 1;
    });
    JSON.stringify(hash).should.be.equal(JSON.stringify({'a': 1}));
  });

  it('remove_copy', function(){
    var ary = [1, 2, 3];
    JSON.stringify(Algos.remove_copy(ary, 3))
      .should.be.equal(JSON.stringify([3]));
  });

  it('remove_copy_if', function(){
    var ary = [1, 2, 3];
    ary = Algos.remove_copy_if(ary, function(e){
      return e < 2;
    });
    JSON.stringify(ary).should.be.equal(JSON.stringify([1]));
    var hash = {'a': 1, 'b': 2};
    hash = Algos.remove_copy_if(hash, function(e){
      return e != 1;
    });
    JSON.stringify(hash).should.be.equal(JSON.stringify({'b': 2}));
  });

  it('replace', function(){
    var ary = [1, 2, 3, 3];
    Algos.replace(ary, 3, 2);
    JSON.stringify(ary).should.be.equal(JSON.stringify([1, 2, 2, 2]));
  });

  it('replace_if', function(){
    var ary = [1, 2, 3];
    Algos.replace_if(ary, function(e){
      return e < 2;
    }, 2);
    JSON.stringify(ary).should.be.equal(JSON.stringify([2, 2, 3]));
    var hash = {'a': 1, 'b': 2};
    Algos.replace_if(hash, function(e){
      return e != 2;
    }, 2);
    JSON.stringify(hash).should.be.equal(JSON.stringify({'a': 2, 'b': 2}));
  });

  it('replace_copy', function(){
    var ary = [1, 2, 3, 3];
    JSON.stringify(Algos.replace_copy(ary, 3, 2))
      .should.be.equal(JSON.stringify([1, 2, 2, 2]));
  });

  it('replace_copy_if', function(){
    var ary = [1, 2, 3];
    ary = Algos.replace_copy_if(ary, function(e){
      return e < 2;
    }, 2);
    JSON.stringify(ary).should.be.equal(JSON.stringify([2, 2, 3]));
    var hash = {'a': 1, 'b': 2};
    hash = Algos.replace_copy_if(hash, function(e){
      return e != 2;
    }, 2);
    JSON.stringify(hash).should.be.equal(JSON.stringify({'a': 2, 'b': 2}));
  });

  it('swap_range', function(){
    var a = [1, 2, 3, 4];
    var b = [4, 3, 2, 1];
    Algos.swap_range(a, b, 1, 3);
    JSON.stringify(a).should.be.equal(JSON.stringify([1, 3, 2, 4]));
    JSON.stringify(b).should.be.equal(JSON.stringify([4, 2, 3, 1]));
  });

  it('reverse', function(){
    var ary = [1, 2, 3];
    Algos.reverse(ary);
    ary.toString().should.be.equal([3, 2, 1].toString());
  });

  it('reverse_copy', function(){
    Algos.reverse_copy(Ary()).toString().
      should.be.equal(Ary().reverse().toString());
  });

  it('shuffle', function(){
    Algos.shuffle(Ary());
  });

  it('rotate', function(){
    var ary = [1, 2, 3, 4];
    Algos.rotate(ary, 2);
    ary.toString().should.be.equal([3, 4, 1, 2].toString());
  });

  it('rotate_copy', function(){
    Algos.rotate_copy(Ary(), 2).toString()
      .should.be.equal([3, 4, 5, 1, 2].toString());
  });

  it('unique', function(){
    var duplicate = [1, 1, 2, 3, 3];
    Algos.unique(duplicate);
    JSON.stringify(duplicate).should.be.equal(JSON.stringify([1, 2, 3]));
  });

  it('unique_copy', function(){
    var duplicate = [1, 1, 2, 3, 3];
    JSON.stringify(Algos.unique_copy(duplicate))
      .should.be.equal(JSON.stringify([1, 2, 3]));
  });
});

File: lua/modifying_sequence_spec.lua
local Algos = require('./modifying_sequence')
local ary = {1, 2, 3, 4, 5}
local table = {a = 1, b = 2}

describe('modifying sequence algorithms', function()
    it('copy', function()
        local dest = {}
        Algos.copy(table, dest)
        assert.are.same(table, dest)
    end)

    it('copyIf', function()
        local dest = {}
        Algos.copyIf(table, dest, function (v)
            return v == 1
        end)
        assert.are.same({a = 1}, dest)
    end)

    it('copyN', function()
        local dest = {}
        Algos.copyN(dest, 3, 3)
        assert.are.same({3, 3, 3}, dest)
    end)

    it('copyBackward', function()
        local dest = {1, 1, 1}
        assert.are.same(0, Algos.copyBackward(ary, dest))

        dest = {1, 1, 1, 1, 1}
        assert.are.same(1, Algos.copyBackward(ary, dest))
        assert.are.same({1, 2, 3, 4, 5}, dest)

        dest = {1, 1, 1, 1, 1, 1}
        assert.are.same(2, Algos.copyBackward(ary, dest))
        assert.are.same({1, 1, 2, 3, 4, 5}, dest)
    end)

    it('fill', function()
        local dest = {1, 1, 1}
        Algos.fill(dest, 3)
        assert.are.same({3, 3, 3}, dest)
    end)

    it('fillN', function()
        local dest = {1, 1, 1}
        Algos.fillN(dest, 2, 3)
        assert.are.same({3, 3, 1}, dest)
    end)

    it('transform', function()
        local ary = {1, 2, 3, 4, 5}
        Algos.transform(ary, function (v)
            return v + 1
        end)
        assert.are.same({2, 3, 4, 5, 6}, ary)
    end)

    it('generate', function()
        local dest = {1, 1, 1}
        Algos.generate(dest, function ()
            return 3
        end)
        assert.are.same({3, 3, 3}, dest)
    end)

    it('generateN', function()
        local dest = {1, 1, 1}
        Algos.generateN(dest, 2, function ()
            return 3
        end)
        assert.are.same({3, 3, 1}, dest)
    end)

    it('remove', function()
        local ary = {1, 2, 3}
        Algos.remove(ary, 3)
        assert.are.same({1, 2}, ary)
    end)

    it('removeIf', function()
        local ary = {1, 2, 3}
        Algos.removeIf(ary, function (k, v)
            return v < 2
        end)
        assert.are.same({2, 3}, ary)
    end)

    it('removeCopy', function()
        local ary = {1, 2, 3}
        assert.are.same({3}, Algos.removeCopy(ary, 3))
    end)

    it('removeCopyIf', function()
        local ary = {1, 2, 3}
        ary = Algos.removeCopyIf(ary, function (k, v)
            return v < 2
        end)
        assert.are.same({1}, ary)
    end)

    it('replace', function()
        local ary = {1, 2, 3, 3}
        Algos.replace(ary, 3, 2)
        assert.are.same({1, 2, 2, 2}, ary)
    end)

    it('replaceIf', function()
        local ary = {1, 2, 3}
        Algos.replaceIf(ary, function (k, v)
            return v < 2
        end, 2)
        assert.are.same({2, 2, 3}, ary)
    end)

    it('replaceCopy', function()
        local ary = {1, 2, 3, 3}
        ary = Algos.replaceCopy(ary, 3, 2)
        assert.are.same({1, 2, 2, 2}, ary)
    end)

    it('replaceCopyIf', function()
        local ary = {1, 2, 3}
        ary = Algos.replaceCopyIf(ary, function (k, v)
            return v < 2
        end, 2)
        assert.are.same({2, 2, 3}, ary)
    end)

    it('swapRange', function()
        local a = {1, 2, 3, 4}
        local b = {4, 3, 2, 1}
        -- unlike other implementation,
        -- the range here is double side closed([start, end])
        Algos.swapRange(a, b, 1, 3)
        assert.are.same({4, 3, 2, 4}, a)
        assert.are.same({1, 2, 3, 1}, b)
    end)

    it('reverse', function()
        local ary = {1, 2, 3}
        Algos.reverse(ary)
        assert.are.same({3, 2, 1}, ary)
    end)

    it('reverseCopy', function()
        local ary1 = {1, 2, 3}
        local ary2 = {3, 2, 1}
        assert.are.same(ary2, Algos.reverseCopy(ary1))
    end)

    it('shuffle', function()
        Algos.shuffle({1, 2, 3})
        Algos.shuffle({1, 2, 3, 4})
    end)

    it('rotate', function()
        local ary = {1, 2, 3, 4}
        Algos.rotate(ary, 3)
        assert.are.same({3, 4, 1, 2}, ary)
    end)

    it('rotateCopy', function()
        assert.are.same({3, 4, 5, 1, 2}, Algos.rotateCopy(ary, 3))
    end)

    it('unique', function()
        local duplicate = {1, 1, 2, 3, 3}
        Algos.unique(duplicate)
        assert.are.same({1, 2, 3}, duplicate)
    end)

    it('uniqueCopy', function()
        local duplicate = {1, 2, 2, 3}
        assert.are.same({1, 2, 3}, Algos.uniqueCopy(duplicate))
    end)
end)
